### DISCUSSION TITLE
[AMD] ICD enumerates fine, but no devices found

### DISCUSSION THREAD

**Discussion by Richardk2n** [pyopencl/pyopencl#794] [Q&A] [answered]
pyopencl finds a platform
```Python
>>> cl.get_platforms()
[<pyopencl.Platform 'AMD Accelerated Parallel Processing' at 0x7f1feeb07f10>]
```
but no devices
```Python
>>> cl.get_platforms()[0].get_devices()
[]
```
Other opencl software works just fine. For example `clinfo`:
```
$ clinfo -l
Platform #0: AMD Accelerated Parallel Processing
 +-- Device #0: gfx906:sramecc+:xnack-
 +-- Device #1: gfx906:sramecc+:xnack-
 +-- Device #2: gfx906:sramecc+:xnack-
 +-- Device #3: gfx906:sramecc+:xnack-
 +-- Device #4: gfx906:sramecc+:xnack-
 +-- Device #5: gfx906:sramecc+:xnack-
 +-- Device #6: gfx90a:sramecc+:xnack-
 `-- Device #7: gfx906:sramecc+:xnack-
```
pyopencl has been installed via pip. I have no rights on the server.

icds are present:
```
$ ls /etc/OpenCL/vendors
amdocl64_60103_sles154_122.icd  amdocl-orca64.icd
```

Any Ideas/Debugging steps?

**Comment by inducer** [comment]
That's super weird, I'm not sure I've seen that failure mode. Some things to try:

- Use `ldd` on `clinfo` and the PyOpenCL shared object to make sure they are both talking to the same `libOpenCL.so`.
- Use `strace` on `clinfo` and device enumeration via PyOpenCL to see how the two paths diverge.

**Reply by Richardk2n** [reply]
`ldd` revealed, that they use vastly different `libOpenCL.so`
`clinfo` uses `/opt/amdgpu-pro/lib64/libOpenCL.so.1`
`_cl.cpython-310-x86_64-linux-gnu.so` uses `.myVenv/lib/python3.10/site-packages/pyopencl/./.libs/libOpenCL-b54a1ea0.so.1.0.0`, which is a lot larger.
Replacing it with the lib from `/opt/` produced an additional error `libstdc++.so.6: version 'GLIBCXX_3.4.30' not found`, because it somehow grabs the wrong `libstadc++.so.6`. It seems like python prepends its own lib directory to the `LD_LIBRARY_PATH`).

My understanding of the `strace` command is practically 0. The output is very long and different from the beginning. Is there anything specific to look for?

**Reply by inducer** [reply, marked as answer]
> My understanding of the strace command is practically 0.

You won't need it. You've already found the root cause.

> Replacing it with the lib from /opt/ produced an additional error libstdc++.so.6: version 'GLIBCXX_3.4.30' not found, because it somehow grabs the wrong libstadc++.so.6.

It should be safe to move that `libstdc++.so.6` aside (or delete it) if you're feeling brave.

**Reply by Richardk2n** [reply]
Thank you, I "replaced" `libstdc++.so.6` using `LD_PRELOAD`because I have no rights on the server to actually modify the file and this worked.

**Comment by kif** [comment]
Could it be a problems of rights to access some devices in `/dev`? 
The `strace` path Andreas suggest should be able to point out which device is the problem.
